### ISSUE TITLE
add logging of custom metrics to eval script

### ISSUE THREAD

**Issue by msalganik** [niznik-dev/predicting-zygosity#18]
For the evaluation script that runs during training, we would like to send custom performance measures to WandB.

For now, let's focus on the case of binary outcome prediction.

~~Here's a list of possible things from scikit learn:
https://scikit-learn.org/stable/api/sklearn.metrics.html~~

For now, let's explore [TorchEval](https://docs.pytorch.org/torcheval/stable/) because it should play well with TorchTune.

![Image](https://github.com/user-attachments/assets/0e519345-8154-43c9-989b-7d2e79050228)

Most useful ones to include first are:
- MeanSquaredError https://docs.pytorch.org/torcheval/stable/generated/torcheval.metrics.MeanSquaredError.html#torcheval.metrics.MeanSquaredError
 
Let's pass these to WandB using exactly the function name from TorchEval. 

In an ideal world, this should be done without have to make any changes to torchtune.

**Comment by msalganik** [comment]
Rather than scikit, maybe we should use pytorch, which has similar functions:
https://docs.pytorch.org/torcheval/main/torcheval.metrics.html

I've never used it before but my guess is that it should be faster and easier, but I don't know.  I'd love to see more about the eval script.

**Comment by niznik-dev** [comment]
Moved the above to #22 , this issue will just focus on logging to WandB

**Comment by niznik-dev** [comment]
After doing some research with Claude, our loss choice in the yaml, CEWithChunkedOutputLoss, is a form of log loss already. Though this isn't my area of expertise so I can't say I completely understand how Cross Entropy loss is equivalent to log loss.

**Comment by msalganik** [comment]
@niznik-dev I did some quick googling.  
- Cross-entropy loss = log loss (https://en.wikipedia.org/wiki/Cross-entropy)
- I'm happy to use CEWithChunkedOutputLoss, which seems to just be a clever/efficient way to calculate cross-entropy loss/log loss.  https://docs.pytorch.org/torchtune/0.6/generated/torchtune.modules.loss.CEWithChunkedOutputLoss.html?highlight=chunked+cross+entropy+loss 
- I've updated the issue to remove the request for log loss

**Comment by msalganik** [comment]
@niznik-dev we didn't get to talk about this at the unconference today.  Please add it to the agenda for the next software meeting.

**Comment by niznik-dev** [comment]
Claude had a good question that I don't know the answer to either... @msalganik ?

"The question is: What exactly do you want MSE to measure?
  - MSE between predicted token IDs and actual token IDs?
  - MSE between predicted probabilities and some target probabilities?
  - Something else?"

**Comment by msalganik** [comment]
@niznik-dev  and looping in @alessandramaranca 

Great question.  When we are doing a task where the output is 0 or 1 (e.g, MZ twin or not) let
- \hat{p}_i be the probability mass on token for 1 for person i(@alessandramaranca knows how to extract this).  \hat{p}_i should always be between 0 and 1 (inclusive but net of numerical fuzz).
- y_i is the actual outcome for person i

Then for some group of people S
MSE_{S} = \frac{\sum_{i \in S} (\hat{p}_i - y_i)^2}{size of S}

Also, since this tool is designed for newcomers, it should have lots of asserts and useful error messages.

For example, there could be an assert that \hat{p}_i should always be between 0 and 1 (inclusive but net of numerical fuzz).  If it is not, then it can throw an error with an informative message.

Tasks where the output is 0 and 1 are not the only thing we will do so please leave to generalize to other kinds of outputs (e.g., output the next events in someone's life).

**Comment by msalganik** [comment]
Also looping in @drigobon. Dan, the place for custom metrics that I mentioned is this pull request.

Given the code that @niznik-dev created, do you think your code getting the embeddings could be added as part of this?

**Comment by msalganik** [comment]
looping in @alessandramaranca might be interested in trying

**Comment by drigobon** [comment]
@msalganik I'd need to work with Mattie to understand exactly what these custom metrics are doing, but in principle we can absolutely add in the embeddings calculation. If the model is already looping over the evaluation data to track metrics like loss (or even the Hessian), then it is implicitly already calculating the relevant embeddings (since they're an intermediate stage to the model outputs). It might even be much better to compute these during the fine-tuning, to avoid re-running the validation data through the model twice...

**Comment by niznik-dev** [comment]
Moving the addition of additional metrics to new tickets...we'll merge this with just the basic outline by itself in a bit.
